core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java

```diff
@@ -487,4 +339,163 @@ public boolean isFinished()
             return current == null && finished;
         }
     }
+
+    private class ChannelFilter
```

Comment:
It would be cleaner for this class to be `static`. When inner class can access anything from the outer one, the contract between them is unclear.
You can pass down a reference to `checkEmpty` via a `Runnable notifyCompletion` or something similar.

Reply:
What is wrong with it being non-static. This is an implementation detail, inner class made to group code better. Passing some Runnables around, on the other hand, seems excessive and overly complicated IMHO.

Reply:
You can pass a `Callable<Void>` if you don't like `Runnable` :)
non-static inner class is not wrong, just that if it is `static` then it is much easier for someone reading the code to understand what is the linkage between the inner and outer class. Otherwise you need to look closer at inner class implementation to see where and how it accesses the fields and methods of the outer class. This also makes it easier to write unit tests for the inner class if we want to do that at some point.